File: doc/code/cpp/elf.cpp
#include <cassert>
#include <cstdint>
#include <iostream>
#include <memory>
#include <sstream>
#include <string>
#include <vector>

#include <LIEF/ELF.hpp>
#include <LIEF/runtime.hpp>

using namespace LIEF::ELF;

void inspect() {
  // lief-doc: inspect-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  std::cout << elf->header().entrypoint();

  for (const LIEF::ELF::Section& section : elf->sections()) {
    std::cout << section.name() << section.content().size() << '\n';
  }
  // lief-doc: inspect-end
}

void modify_write() {
  // lief-doc: write-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  elf->add_library("libdemo.so");
  elf->write("new.elf");
  // lief-doc: write-end
}

void write_bytes() {
  // lief-doc: write-bytes-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  std::ostringstream os;
  elf->write(os);
  std::string buffer = os.str();

  const auto* start = reinterpret_cast<const uint8_t*>(buffer.data());
  size_t size = buffer.size();
  // lief-doc: write-bytes-end
  (void)start;
  (void)size;
}

void add_loaded_segment() {
  // lief-doc: add-segment-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  LIEF::ELF::Segment segment;
  segment.type(LIEF::ELF::Segment::TYPE::LOAD);
  segment.content({1, 2, 3});

  LIEF::ELF::Segment* new_segment = elf->add(segment);
  elf->write("new.elf");
  // lief-doc: add-segment-end
  (void)new_segment;
}

void add_loaded_section() {
  // lief-doc: add-section-loaded-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  LIEF::ELF::Section section(".lief_demo");
  section.content({1, 2, 3});

  LIEF::ELF::Section* new_section = elf->add(section, /*loaded=*/true);
  elf->write("new.elf");
  // lief-doc: add-section-loaded-end
  (void)new_section;
}

void add_unloaded_section() {
  // lief-doc: add-section-unloaded-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  LIEF::ELF::Section section(".metadata");
  section.content({1, 2, 3});

  LIEF::ELF::Section* new_section = elf->add(section, /*loaded=*/false);
  elf->write("new.elf");
  // lief-doc: add-section-unloaded-end
  (void)new_section;
}

void parse_from_dump() {
  // lief-doc: dump-start
  auto elf = LIEF::ELF::Parser::parse_from_dump("module.dump", 0x7f9b98e00000);

  for (const LIEF::ELF::Segment& segment : elf->segments()) {
    std::cout << to_string(segment.type()) << '\n';
  }
  // lief-doc: dump-end
}

void module_dump() {
  // lief-doc: dump-runtime-start
  // Find the module to dump in the current process
  auto mod = LIEF::runtime::module_from_name("libc.so.6");

  // Dump the module's memory into a file (the raw bytes are also returned)
  std::vector<uint8_t> data = mod->dump("module.dump");

  auto elf = LIEF::ELF::Parser::parse_from_dump("module.dump", mod->imagebase());
  // lief-doc: dump-runtime-end
  (void)data;
  (void)elf;
}

void advanced_parse_write() {
  // lief-doc: advanced-start
  LIEF::ELF::ParserConfig parser_config;
  parser_config.parse_overlay = false;

  auto elf = LIEF::ELF::Parser::parse("my.elf", parser_config);

  LIEF::ELF::Builder::config_t builder_config;
  builder_config.gnu_hash = false;

  elf->write("new.elf", builder_config);
  // lief-doc: advanced-end
}

void rpath_add() {
  // lief-doc: rpath-add-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  LIEF::ELF::DynamicEntryRunPath runpath("$ORIGIN:/opt/lib64");

  elf->add(runpath);

  LIEF::ELF::DynamicEntryRunPath other_runpath(
      std::vector<std::string>{"$ORIGIN", "/opt/lib64"}
  );

  elf->add(other_runpath);

  elf->write("updated.elf");
  // lief-doc: rpath-add-end
}

void rpath_change() {
  // lief-doc: rpath-change-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  auto* runpath = elf->get(LIEF::ELF::DynamicEntry::TAG::RUNPATH)
                      ->cast<LIEF::ELF::DynamicEntryRunPath>();

  assert(runpath != nullptr);

  runpath->runpath("$ORIGIN:/opt/lib64");
  runpath->append("lib-x86_64-gnu");

  elf->write("updated.elf");
  // lief-doc: rpath-change-end
}

void rpath_remove() {
  // lief-doc: rpath-remove-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  // Remove **all** DT_RUNPATH entries
  elf->remove(LIEF::ELF::DynamicEntry::TAG::RUNPATH);

  // Remove all entries that contain '$ORIGIN'
  std::vector<LIEF::ELF::DynamicEntryRunPath*> to_remove;
  for (DynamicEntry& entry : elf->dynamic_entries()) {
    if (auto* dt_entry = entry.cast<LIEF::ELF::DynamicEntryRunPath>()) {
      if (dt_entry->runpath().find("$ORIGIN") != std::string::npos) {
        to_remove.push_back(dt_entry);
      }
    }
  }

  for (LIEF::ELF::DynamicEntryRunPath* entry : to_remove) {
    elf->remove(*entry);
  }

  elf->write("updated.elf");
  // lief-doc: rpath-remove-end
}

void symver_remove_symbol() {
  // lief-doc: symver-symbol-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  LIEF::ELF::Symbol* sym = elf->get_dynamic_symbol("printf");

  assert(sym != nullptr);

  sym->symbol_version()->as_global();

  elf->write("updated.elf");
  // lief-doc: symver-symbol-end
}

void symver_remove_library() {
  // lief-doc: symver-library-start
  std::unique_ptr<LIEF::ELF::Binary> elf;

  elf->remove_version_requirement("libm.so.6");

  elf->write("updated.elf");
  // lief-doc: symver-library-end
}

void parse_forms() {
  // lief-doc: parse-start

  // Using a file path as a std::string
  std::unique_ptr<LIEF::ELF::Binary> elf = LIEF::ELF::Parser::parse("/bin/ls");

  // Using a vector
  std::vector<uint8_t> my_raw_elf;
  elf = LIEF::ELF::Parser::parse(my_raw_elf);
  // lief-doc: parse-end
}

File: doc/code/rust/src/elf.rs
use std::path::PathBuf;

use lief::elf::dynamic;
use lief::generic::Section;

pub fn inspect(some_elf: &lief::elf::Binary) {
    // lief-doc: inspect-start
    let elf: &lief::elf::Binary = some_elf;

    println!("{}", elf.header().entrypoint());

    for section in elf.sections() {
        println!("{} {}", section.name(), section.content().len());
    }
    // lief-doc: inspect-end
}

pub fn modify_write(some_elf: &mut lief::elf::Binary) {
    // lief-doc: write-start
    let elf: &mut lief::elf::Binary = some_elf;

    elf.add_library("libdemo.so");
    elf.write("new.elf");
    // lief-doc: write-end
}

pub fn write_bytes(some_elf: &mut lief::elf::Binary) {
    // lief-doc: write-bytes-start
    let elf: &mut lief::elf::Binary = some_elf;

    let bytes: Vec<u8> = elf.write_to_bytes();
    // lief-doc: write-bytes-end
}

pub fn add_loaded_segment(some_elf: &mut lief::elf::Binary) {
    // lief-doc: add-segment-start
    let elf: &mut lief::elf::Binary = some_elf;

    let mut segment = lief::elf::Segment::new();
    segment.set_type(lief::elf::segment::Type::LOAD);
    segment.set_content(&[1, 2, 3]);

    elf.add_segment(&segment);
    elf.write("new.elf");
    // lief-doc: add-segment-end
}

pub fn add_loaded_section(some_elf: &mut lief::elf::Binary) {
    // lief-doc: add-section-loaded-start
    let elf: &mut lief::elf::Binary = some_elf;

    let section = lief::elf::Section::new_with_content(".lief_demo", &[1, 2, 3]);

    elf.add_section(
        &section,
        /* loaded= */ true,
        lief::elf::binary::SecInsertPos::AUTO,
    );
    elf.write("new.elf");
    // lief-doc: add-section-loaded-end
}

pub fn add_unloaded_section(some_elf: &mut lief::elf::Binary) {
    // lief-doc: add-section-unloaded-start
    let elf: &mut lief::elf::Binary = some_elf;

    let section = lief::elf::Section::new_with_content(".metadata", b"version: 1.2.3");

    // /!\ Note that loaded is set to false here
    // -----------------------------------------
    elf.add_section(
        &section,
        /* loaded= */ false,
        lief::elf::binary::SecInsertPos::AUTO,
    );
    elf.write("new.elf");
    // lief-doc: add-section-unloaded-end
}

pub fn parse_from_dump() {
    // lief-doc: dump-start
    let elf = lief::elf::Binary::parse_from_dump("module.dump", 0x7f9b_98e0_0000).unwrap();

    for segment in elf.segments() {
        println!("{:?} {:#x}", segment.p_type(), segment.virtual_address());
    }
    // lief-doc: dump-end
}

pub fn module_dump() {
    // lief-doc: dump-runtime-start
    use lief::runtime::Module;

    let module = lief::runtime::module_from_name("libc.so.6").unwrap();

    // Dump the module's memory into a file (the raw bytes are also returned)
    let data = module.dump_to_file("module.dump");

    let elf = lief::elf::Binary::parse_from_dump("module.dump", module.imagebase()).unwrap();
    // lief-doc: dump-runtime-end
}

pub fn advanced_parse_write() {
    // lief-doc: advanced-start
    let mut parser_config = lief::elf::ParserConfig::default();
    parser_config.parse_overlay = false;

    let mut elf = lief::elf::parse_with_config("my.elf", &parser_config).unwrap();

    let mut builder_config = lief::elf::builder::Config::default();
    builder_config.gnu_hash = false;

    elf.write_with_config("new.elf", builder_config);
    // lief-doc: advanced-end
}

pub fn rpath_add(some_elf: &mut lief::elf::Binary) {
    // lief-doc: rpath-add-start
    let elf: &mut lief::elf::Binary = some_elf;

    let runpath = lief::elf::dynamic::RunPath::new("$ORIGIN:/opt/lib64");

    elf.add_dynamic_entry(&runpath);

    let other_runpath = lief::elf::dynamic::RunPath::with_paths(&["$ORIGIN", "/opt/lib64"]);

    elf.add_dynamic_entry(&other_runpath);

    let output = PathBuf::from("updated.elf");

    elf.write(output.as_path());
    // lief-doc: rpath-add-end
}

pub fn rpath_change(some_elf: &mut lief::elf::Binary) {
    // lief-doc: rpath-change-start
    let elf: &mut lief::elf::Binary = some_elf;

    if let Some(dynamic::Entries::RunPath(mut runpath)) =
        elf.dynamic_entry_by_tag(dynamic::Tag::RUNPATH)
    {
        runpath.set_runpath("$ORIGIN:/opt/lib64");
        runpath.append("lib-x86_64-gnu");
    }

    let output = PathBuf::from("updated.elf");

    elf.write(output.as_path());
    // lief-doc: rpath-change-end
}

pub fn rpath_remove(some_elf: &mut lief::elf::Binary) {
    // lief-doc: rpath-remove-start
    let elf: &mut lief::elf::Binary = some_elf;

    // Remove **all** DT_RUNPATH entries
    elf.remove_dynamic_entries_by_tag(dynamic::Tag::RUNPATH);

    // Remove all entries that contain '$ORIGIN'
    elf.remove_dynamic_entry_if(|e| {
        if let dynamic::Entries::RunPath(runpath) = e {
            return runpath.runpath().contains("$ORIGIN");
        }
        false
    });

    let output = PathBuf::from("updated.elf");
    elf.write(output.as_path());
    // lief-doc: rpath-remove-end
}

pub fn symver_remove_symbol(some_elf: &mut lief::elf::Binary) {
    // lief-doc: symver-symbol-start
    let elf: &mut lief::elf::Binary = some_elf;

    if let Some(sym) = elf.dynamic_symbol_by_name("printf")
        && let Some(mut symver) = sym.symbol_version()
    {
        symver.as_global();
    }

    let output = PathBuf::from("updated.elf");
    elf.write(output.as_path());
    // lief-doc: symver-symbol-end
}

pub fn symver_remove_library(some_elf: &mut lief::elf::Binary) {
    // lief-doc: symver-library-start
    let elf: &mut lief::elf::Binary = some_elf;

    elf.remove_version_requirement("libm.so.6");

    let output = PathBuf::from("updated.elf");
    elf.write(output.as_path());
    // lief-doc: symver-library-end
}

pub fn parse_forms() {
    // lief-doc: parse-start
    let elf: lief::elf::Binary = lief::elf::Binary::parse("/bin/ls").unwrap();
    // lief-doc: parse-end
}
